### PR TITLE
Http query parsers

### PR DESCRIPTION
**Pull request by wilsoniya**
This PR introduces reusable query parameter parsers, specifically:

- ParseCoordinates
- ParseTime

@briancharous @greg-mclain just CCing you in on this PR as tools users

### CODE REVIEW

File: http/parsers.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	latStr := r.Form.Get(latFieldName)
+	lonStr := r.Form.Get(lonFieldName)
+	latPresent, lonPresent := false, false
+	var lat, lon float64
+
+	if latStr != "" {
+		// case: some value supplied for latitude
+		_lat, errLat := strconv.ParseFloat(latStr, 64)
+		if errLat != nil {
+			return nil, fmt.Errorf("unable to parse %v", latFieldName)
```
Comment (benjigoldberg) [inline http/parsers.go]: We should likely wrap the error instead with [`xerrors`: `xerrors.Errorf`](https://godoc.org/golang.org/x/xerrors)

This applies for a number of errors below, I'll just make the comment once

File: http/parsers.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+	if lat < -90 || lat > 90 {
+		return nil, fmt.Errorf("%v must be in [-90, 90]", latFieldName)
+	}
+
+	if lon < -180 || lon > 180 {
+		return nil, fmt.Errorf("%v must be in [-180, 180]", lonFieldName)
+	}
+
+	if latPresent && lonPresent {
```
Comment (benjigoldberg) [inline http/parsers.go]: This is a minor idiomatic suggestion here, as suggested in [Effective Go](https://golang.org/doc/effective_go.html#control-structures), but we tend to prefer when conditionals are used to encapsulate degenerate cases only. Eg, the function should end in the "happy-path" return value if possible, eg:

```
if latPresent != lonPresent {
    return nil, fmt.Errorf("both %v and %v must be provided", latFieldName, lonFieldName)
} else if !latPresent && !lonPresent {
    return nil, nil
}
return &Coordinates{lat, lon}, nil
```
Additionally, is it not an error if no lat and lon is found?

File: http/parsers.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+// ParseCoordinates reads and parses from the query parameters to the supplied
+// request latitude and logitude corresponding to latFieldName and
+// lonFieldName, respectively. An error is returned if only one of the named
+// fields is present, if either value cannot be parsed as a float, or if either
+// value is out of range for decimal latitude and longitude. The returned
+// struct reference is nil if none of latFieldName or lonFieldName are present
+// in the query parameters to the given request.
+func ParseCoordinates(r *http.Request, latFieldName, lonFieldName string) (*Coordinates, error) {
+	r.ParseForm()
```
Comment (benjigoldberg) [inline http/parsers.go]: I wonder if parseform is a noop if its already been called on a request 🤔

File: http/parsers.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+// ParseTime reads and parses from the query parameters to the supplied request
+// a Time value corresponding to fieldName. An error is returned if a Time
+// could not be parsed from the given field. The zero-valued Time is returned
+// if the given field is not present in the query parameters to the supplied
+// request.
+func ParseTime(r *http.Request, fieldName string) (time.Time, error) {
+	r.ParseForm()
+	fieldStr := r.Form.Get(fieldName)
+	if fieldStr != "" {
```
Comment (benjigoldberg) [inline http/parsers.go]: I would argue this should instead be
```
if fieldStr == "" {
    return time.Time{}, nil
}
... <rest of logic from if statement goes here>

File: http/parsers.go (hunk 3 above)
Comment (wilsoniya) [reply http/parsers.go]: Not sure if it's a noop, but it is explicitly called out as idempotent.

Comment (wilsoniya) [reply http/parsers.go]: Sidebar, IMO it's really shitty API design forcing the consumer to call a method for its side effects in this way.

File: http/parsers.go (hunk 2 above)
Comment (wilsoniya) [reply http/parsers.go]: Thanks for helping me w/ go idioms. 

> Additionally, is it not an error if no lat and lon is found?

I designed these parsers to *optionally* parse values from query strings, i.e., if the sought values are not present that's not considered an error. The rationale is that query params are often used to provide optional information e.g. for filters and sorting. IMO a missing value does not express an error, and furthermore, disambiguating an error indicating a missing value from any other error would make these APIs more difficult to use. 

Unfortunately Go does not make it easy to express "missing" values across interfaces. Where possible I opt to return zero-valued structs where that carries unambiguous meaning, elsewhere I'm opting to return nillable pointers (which is very unfortunate).

File: http/parsers.go (hunk 4 above)
Comment (wilsoniya) [reply http/parsers.go]: Nice :+1:

File: http/parsers.go (hunk 2 above)
Comment (benjigoldberg) [reply http/parsers.go]: ah, I see, ok.

File: http/parsers.go (hunk 3 above)
Comment (benjigoldberg) [reply http/parsers.go]: yes, I agree

**Pull request merged**